### PR TITLE
WIP IA-2370 Disable user

### PR DESCRIPTION
**Pull request by Qi77Qi**
Ticket: https://broadworkbench.atlassian.net/browse/IA-2370

Before the break, we decided on the approach where leo will publish the bad user and sam will subscribe to the topic

Depends on https://github.com/broadinstitute/firecloud-develop/pull/2372

Test:
Manually sent a message to `projects/broad-dsde-dev/topics/terra-cryptomining` topic, and sam receives the message properly.

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: src/main/resources/reference.conf [hunk 1]
```diff
@@ -792,8 +792,8 @@ resourceTypes = {
 
 
 distributedLock {
-  retryInterval: 5s
-  maxRetry: 10
+  retryInterval = 5s
```
Comment (Qi77Qi) [inline src/main/resources/reference.conf:795]: was this working before? 🤔

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala [hunk 2]
```diff
@@ -81,11 +84,11 @@ object Boot extends IOApp with LazyLogging {
 
         _ <- dependencies.cloudExtensionsInitializer.onBoot(dependencies.samApplication)
 
-        binding <- IO.fromFuture(IO(Http().newServerAt("0.0.0.0", 8080).bind(dependencies.samRoutes.route))).onError {
+        httpServer = IO.fromFuture(IO(Http().newServerAt("0.0.0.0", 8080).bind(dependencies.samRoutes.route))).onError {
           case t: Throwable => IO(logger.error("FATAL - failure starting http server", t)) *> IO.raiseError(t)
-        }
-        _ <- IO.fromFuture(IO(binding.whenTerminated))
-        _ <- IO(system.terminate())
+        }.void
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:89]:
```
  /*
   * Ignores the result of this IO.
   */
  def void
```

(I'm a bit confused) isn't httpServer just IO[Unit]? Why would that be useful to pass into the next line?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala [hunk 3]
```diff
@@ -81,11 +84,11 @@ object Boot extends IOApp with LazyLogging {
...
+        allStreams = fs2.Stream.eval(httpServer) :: dependencies.cryptominingSubscriberProcess
+        _ <- fs2.Stream.emits(allStreams).covary[IO].parJoin(allStreams.length).compile.drain
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:91]: not knowing a lot about streams, could you summarize what is happening here? (and maybe I should read up on Streams)

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 4]
```diff
@@ -507,11 +507,9 @@ class ResourceService(
           // resources with auth domains logically can't have public policies but also technically allowing them poses a problem
           // because the logic for public resources is different. However, sharing with the auth domain should have the
           // exact same effect as making a policy public: anyone in the auth domain can access.
-          if (public)
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: why was this removed? based on the function description seems useful to have

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import fs2.Stream
+import io.chrisdavenport.log4cats.StructuredLogger
+import io.circe.Decoder
+import org.broadinstitute.dsde.workbench.google2.{Event, GoogleSubscriber}
+import org.broadinstitute.dsde.workbench.model.WorkbenchUserId
+import org.broadinstitute.dsde.workbench.sam.service.UserService
+import org.broadinstitute.dsde.workbench.sam.util.SamRequestContext
+
+// Each instance of Sam will run a subscriber, since they all subscribe to the same topic via same subscription,
+// messages will be distributed among subscribers instead of distributed to all of them, which is what we want.
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala:14]: thanks for the explanation 👍

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      // To conditionally acking message, we'll need to config a dead letter topic where messages reach max retry can go to,
+      // which we're not doing for this use case.
+    _ <- IO{event.consumer.ack()}
+    } yield ()
+  }
+
+  private def messageResponder(
+                                         message: CryptominingUserMessage
+                                       ): IO[Unit] = IO.fromFuture(IO(userService.disableUser(message.userSubjectId, SamRequestContext(None)))).void
+}
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala:40]: is it possible to make a unit test for this? secondly, would it be worth it?

**Review by andy7i — COMMENTED** [review]
comments / questions

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala (hunk 3 above)
Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:91]: This is a way to run various processes concurrently. In this case, it's the cryptomining subscriber, and the main sam http server.

For your question above, because the Streams here are expected to be all `Stream[IO, Unit]`, hence we're using `void` up there. Without `void`, it'll return `IO[Http.ServerBinding]`, which can't be combined with other Streams for running concurrently.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import io.chrisdavenport.log4cats.StructuredLogger
+import io.circe.Decoder
+import org.broadinstitute.dsde.workbench.google2.{Event, GoogleSubscriber}
+import org.broadinstitute.dsde.workbench.model.WorkbenchUserId
+import org.broadinstitute.dsde.workbench.sam.service.UserService
+import org.broadinstitute.dsde.workbench.sam.util.SamRequestContext
+
+// Each instance of Sam will run a subscriber, since they all subscribe to the same topic via same subscription,
+// messages will be distributed among subscribers instead of distributed to all of them, which is what we want.
+class CryptominingSubscriber(subscriber: GoogleSubscriber[IO, CryptominingUserMessage],
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala:15]: I don't think we should mention cryptomining. This is a general purpose mechanism for any automated external process to tell sam to disable a user. Cryptomining is just the first use of this.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala [hunk 8]
```diff
@@ -133,7 +136,23 @@ object Boot extends IOApp with LazyLogging {
...
+      }
+    } yield createAppDepenciesWithSamRoutes(appConfig, cloudExtensionsInitializer, userService, cryptominingSubscriberProcess, foregroundAccessPolicyDAO, foregroundDirectoryDAO, identityConcentrator)
+
+  private def initCryptominingSubscriber(cryptominingSubscriberConfig: SubscriberConfig, userService: UserService)
+                                        (implicit logger: StructuredLogger[IO]): effect.Resource[IO, List[fs2.Stream[IO, Unit]]] = for {
+    subscriberQueue <- cats.effect.Resource.liftF(InspectableQueue.bounded[IO, Event[CryptominingUserMessage]](200))
+    cryptominingSubscriber <- GoogleSubscriber.resource[IO, CryptominingUserMessage](cryptominingSubscriberConfig, subscriberQueue)
+  } yield {
+    val subscriber = new CryptominingSubscriber(cryptominingSubscriber, userService)
+    List(subscriber.process, fs2.Stream.eval(cryptominingSubscriber.start))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:154]: can you explain why this is a list of streams?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala [hunk 9]
```diff
@@ -133,7 +136,23 @@ object Boot extends IOApp with LazyLogging {
...
+
+      cryptominingSubscriberProcess <- appConfig.googleConfig match {
+        case Some(value) => initCryptominingSubscriber(value.googleServicesConfig.cryptominingSubscriber, userService)
+        case None => effect.Resource.liftF[IO, List[fs2.Stream[IO, Unit]]](IO(List.empty))
+      }
+    } yield createAppDepenciesWithSamRoutes(appConfig, cloudExtensionsInitializer, userService, cryptominingSubscriberProcess, foregroundAccessPolicyDAO, foregroundDirectoryDAO, identityConcentrator)
+
+  private def initCryptominingSubscriber(cryptominingSubscriberConfig: SubscriberConfig, userService: UserService)
+                                        (implicit logger: StructuredLogger[IO]): effect.Resource[IO, List[fs2.Stream[IO, Unit]]] = for {
+    subscriberQueue <- cats.effect.Resource.liftF(InspectableQueue.bounded[IO, Event[CryptominingUserMessage]](200))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:150]: why 200?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+                             userService: UserService
+                            )(implicit logger: StructuredLogger[IO], cs: ContextShift[IO]) {
+  val process: Stream[IO, Unit] = subscriber.messages
+    .parEvalMapUnordered(2)(messageHandler) // concurrency is low since we don't need high performance in disabling cryptomining users.
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala]: why not `evalMap` and forget the question of concurrency? There are other sam instances too so this is never single threaded anyway.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    for {
+      res <- messageResponder(event.msg).attempt
+      _ <- res match {
+        case Left(e)  => logger.error(e)("Fail to process pubsub message")
+        case Right(_) => IO.unit
+      }
+      // acking the message whether or not we successfully processed the message since this is not essential part of
+      // terra functionality. That is, if we didn't disable a bad user, it's no big deal.
+      // To conditionally acking message, we'll need to config a dead letter topic where messages reach max retry can go to,
+      // which we're not doing for this use case.
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala:32]: I disagree with this. Maybe for cryptomining it is not a big deal but I think the general contract for sam should be stronger. I think we should only ack on success.

**Review by dvoet — CHANGES_REQUESTED** [review]
There are a few detailed comments. In general though, this is a departure from how most of the sam code works. Since this team needs to maintain it I am going to discuss it with them.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala (hunk 4 above)
Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: good catch! this is probably from bad merge

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala (hunk 7 above)
Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala:15]: hmm..think my perspective is a bit different...this is a subscriber specifically for `terra-cryptomining` topic (which I was assuming we can potentially have other messages related to cryptomining activity other than disabling users), we can't really use this for another topic if there's another use case for disabling users.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala (hunk 10 above)
Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala]: I made it 2 just so if in the future there's need to increase concurrency, people don't have to figure out how to use `parEvalMapUnordered`...happy to change this to no concurrency

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala (hunk 11 above)
Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala:32]: this ties back to https://github.com/broadinstitute/sam/pull/494/files/75635f99381bdba2035a10e72098ac34e11fd052#r552697473...

The reason why I'm choosing to ack blindly here is because if we choose to nack, then we should set up max retry, which means we should set up a deadlettertopic for receiving messages that have exceeded max retry, otherwise, nacked messages will be resent forever (I'm not certain if it's forever or there's a default max at some point, but from what we see in leo, it looked like forever). I didn't want to introduce more complexity if not necessary, but if you think it's worth it, I'm happy to add those

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala (hunk 6 above)
Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala:40]: what is the test in your mind trying to validate? if it's just disabling user, I think it should be covered in userService's unit test

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala (hunk 8 above)
Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:154]: `cryptominingSubscriber.start` is for starting up the underlying google subscriber, `subscriber.process` is for actually running the stream that processes messages that the subscriber receives

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala (hunk 9 above)
Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:150]: it's just best guess to start. It's the number of messages that the subscriber will queue up before back pressure happens.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala (hunk 3 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala:91]: OH ok. I think I also missed the `eval` in `fs2.Stream.eval(httpServer)` -- which would allow the stream to evaluate the httpServer IO.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala (hunk 6 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala:40]: that makes sense. Would it be useful to test this subscriber? (I don't know how we would test it either, but just wanted to check and ask to understand why we would or wouldn't want to test this)

Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala:40]: anything specific you want to test about the subscriber? (asking in case I'm missing something)

Subscriber has unit test in google2 to make sure it can receive message properly.
The only thing worth testing here is the actual handler, but the handler here is quite simple cuz it's just calling userSerivce.disableUser, which I assume is covered by userSerivce's unit test

File: src/main/resources/reference.conf (hunk 1 above)
Comment (gpolumbo-broad) [reply src/main/resources/reference.conf:795]: https://github.com/lightbend/config/blob/master/HOCON.md#key-value-separator

Yes

**Review by gpolumbo-broad — CHANGES_REQUESTED** [review]
Hi Qi, we discussed this PR pretty extensively as a team and we are apprehensive about introducing further usage of cats or fs2 unless it is clearly justified.  This was not an easy decision for us to make.  We want to learn and use new technology, but paramount to that we want the code to be easily readable and maintainable for engineers of all seniority levels.  Our experience has shown us a steep learning curve with these libraries and the pure FP principles behind them.  Given these points and the direction of the organization towards more Java work, we would prefer to not introduce the usage of fs2 streams at this time and instead implement this feature using Akka Actors so that it is in alignment with existing code in Sam.

We are very appreciative that you took the time to implement this feature for us and we know that this feature is important to the Leo team and the health of the platform in general.  Our team can take on the reimplementation of this feature using Akka Actors, unless you or your team is particularly keen on doing it, in which case you are welcome to take it on.   Please let us know how you would like to proceed so that we can prioritize this work accordingly.

On a personal level I want to acknowledge your expertise in FP and the training and guidance you are providing to all of us.  You are clearly a leader in this space and I hope that you understand that even though our team is requesting this change, we all appreciate your FP knowledge and expertise.

**Comment by Qi77Qi** [comment]
@gpolumbo-broad thanks for the reply. in that case, I'll leave the work to you guys...
